File: apple/CabalmailKit/Sources/CabalmailKit/Compose/SignatureFormatter.swift
import Foundation

/// Pure value-type helper that inserts a plain-text signature into a compose
/// body.
///
/// Kept out of `ComposeViewModel` so the string-munging is independent of
/// SwiftUI / `@Observable` and can be exercised directly from
/// `CabalmailKitTests`. The signature is always introduced with the RFC 3676
/// signature delimiter `"-- "` (dash-dash-space) on its own line, so
/// downstream mail clients can collapse / strip the signature block the
/// same way every UNIX mail client has since Pine.
public enum SignatureFormatter {
    /// Produces the initial body for a new compose.
    ///
    /// Matches the layout the user expects for each entry point:
    ///
    /// - **New message** (empty base). Result is `\n<delim><signature>`, so
    ///   the cursor lands on the blank line above the signature and the
    ///   user can type above it.
    /// - **Reply / forward** (base leading with `\n\n` for attribution +
    ///   quoted original). Result is `<delim><signature><base>`, placing
    ///   the signature on its own line immediately before the attribution.
    /// - **Any other base** (legacy drafts or tests). Result is
    ///   `<delim><signature>\n<base>`.
    ///
    /// Passing an empty `signature` is a no-op — the original base is
    /// returned unchanged, including the empty-string case.
    public static func seedBody(base: String, signature: String) -> String {
        guard !signature.isEmpty else { return base }
        let signatureBlock = "\n-- \n" + signature
        if base.isEmpty {
            return "\n" + signatureBlock
        }
        if base.hasPrefix("\n\n") {
            return signatureBlock + base
        }
        return signatureBlock + "\n" + base
    }
}

File: android/kit/src/main/kotlin/com/cabalmail/kit/compose/SignatureFormatter.kt
package com.cabalmail.kit.compose

/**
 * Inserts the plain-text signature into a compose seed, always introduced
 * by the RFC 3676 delimiter (`-- ` on its own line) so downstream clients
 * can collapse it. Mirrors the Apple `SignatureFormatter`:
 *
 * - new message (empty base): `\n-- \n<sig>` — cursor lands above it;
 * - reply / forward (base leads with `\n\n`): `-- \n<sig>` + base — the
 *   signature sits right before the attribution;
 * - anything else: `-- \n<sig>\n` + base.
 *
 * An empty signature is a no-op.
 */
object SignatureFormatter {
    fun seedBody(
        base: String,
        signature: String,
    ): String {
        if (signature.isEmpty()) {
            return base
        }
        val block = "\n-- \n$signature"
        return when {
            base.isEmpty() -> "\n$block"
            base.startsWith("\n\n") -> block + base
            else -> "$block\n$base"
        }
    }
}
